docs/eclipse.1
.TH Eclipse 1 "(c) 2001 Iban HATCHONDO"
.\"$Id: eclipse.1,v 1.15 2004/02/12 23:30:23 ihatchondo Exp $

.SH NAME
eclipse - a window manager in Common Lisp


.SH SYNOPSIS
\fB eclipse \fP[ OPTIONS ]


.SH DESCRIPTION
\fIeclipse\fP is a window manager for the X Window System. As a
window manager it provides most standard window managing features,
such as titlebar, icon management, keyboard shortcuts, and a powerful
interaction mode.

This program in usually started by the user's session manager or
startup script. When used from \fIxdm(1)\fP or \fIxinit(1)\fP without
a session manager, \fIeclipse\fP is frequently executed in the
foreground as the last client. When run this way eclipse causes the
session to be terminated. Otherwise the install adds a gnome entry in
the window manager section. To use eclipse under a gnome-session, use
the traditional way with the gnome-control-center (\fIgnomecc\fP).

In most cases, application windows are surrounded by a "frame" with
titlebar at the top and a special border around. The titlebar contains
the name of the application window and some buttons near the right
edge. Those buttons let you close, (un)maximize, and iconify the
window (in order form the right-most to the left-most). 

For the (un)maximization button, pressing pointer Button1 (usually the
left-most button unless it has been changed with \fIxmodmap\fP)
(un)maximizes the window in both directions (i.e., vertically and
horizontally). Pressing pointer Button2 (middle one) (un)maximizes
vertically, and of course pressing pointer Button3 (right-most)
(un)maximizes horizontally.

Windows are re-sized by pressing any pointer button on the window
border, dragging the pointer over the edge that is to be moved, and
releasing the pointer when desired. During this operation the window
is represented as a 3x3 grid that shows you the future window
geometry.
Similarly, windows are moved by pressing any pointer button on the
titlebar. Two modes are available : opaque or box (3x3 grid).

When new windows are created, \fIeclipse\fP will honor any size and
location information requested by the user (usually through the
\fI-geometry\fP command line argument or resources for the individual
applications). 

Some predefined keyboard shortcuts also exist (that are user
re-definable in the config file. See the keystroke section).

.TP
.B\-Alt-Tab\fP and \fBAlt-A\fP 
circulate window up and down.
.TP
.B\-Ctrl-Leftarrow\fP and \fBCtrl-Rightarrow\fP 
go to the previous or next workspace.
.TP
.B\-Ctrl-Alt-arrow\ 
move the pointer up, down, right, or left, depending of the "arrow"
key you pressed.
.TP
.B\-Ctrl-Alt-( z, w, x )\ 
emulate button-press. Use \fBz\fP for button1 and so on with \fBw\fP
\fBx\fP. This emulation does not work for re-sizing or moving windows,
nor does it work on the root-window.


.SH OPTIONS
.TP
\fB--display\fP=display_specification 
starts eclipse for managing the given X display.
.TP
\fB--sm-client-id\fP=ID
specify a session management ID.
.TP
\fB--activate-log\fP
specifies that errors must be logged in a file. The log file will be
named as follows: eclipse-YYYY-MM-DD.log


.SH MENUS
\fIeclipse\fP also provides three pop-up menus when pressing pointer
buttons on the root-window.

\-button1\ : a customizable application menu, that permits you to
launch your favorite applications.

\-button2\ : a recapitulation of the current running applications
(even if iconified) sorted by workspace. Selecting any application is
this menu will change the stacking order of that application to be on
top of the others and will change your current desktop if necessary. 

\-button3\ : four functions for managing windows : move, re-size, close,
destroy.


.SH CUSTOMIZATION
Two way of customizing \fIeclipse\fP exist.

The first one is by using the configuration file named
\fB$HOME/.eclipse\fP. If no local file can be found, \fIeclipse\fP will
use the built-in default file that can be found in the install
directory. \fBThis configuration file is written in Common
Lisp\fP. Future improvements may add a graphical user interface for
this file.

The second way to customize the window manager is by using the
\fBtelnet\fP server for it. (I guess it probably only works for CMUCL
at this time). To use it, do a \fBtelnet\fP on \fBport 6789\fP. The
password is \fB"clara"\fP (with the double quotes). I know there is a
problem of security with the port number and the password, but it will
be fixed in the future. 
The telnet will drop you in a lisp remote session that share its
environment with the Lisp session that is running the window
manager. Of course your modification will not be saved at the end of
the session, but you can \fBeasily\fP test some modifications that you
want to add in your config file, change in the code, or that you want
to send me for correcting bugs and/or add new features.

.SH VARIABLES & FUNCTIONS
Here are some variables that can be changed (either in the config file
or in the telnet session) :

.TP
.B\-*change-desktop-message-active-p*\ \fIboolean\fP (or t nil) 
If you want to be notified when the desktop changes this value should
be \fIt\fP. In that case, a change will display a small window with
the number or the name (if sets)  of the new desktop. If the value is
\fInil\fP, nothing will happen.
.TP
.B\-*save-and-restore-pointer-position-during-workspace-switch*\
\fIboolean\fP (or t nil)
If you want the mouse pointer position to be saved and restore during
workspace switching then set it to \fIt\fP. Default value is \fInil\fP.
.TP
.B\-*move-mode*\ \fIkeyword\fP (or \fB:box :opaque\fP) 
As usual :box will draw a 3x3 grid to show you the window move ;
:opaque moves the window directly. (The box mode looks good on
laptops).
.TP
.B\-*resize-mode*\ \fIkeyword\fP (or \fB:box :opaque\fP) 
As usual :box will draw a 3x3 grid to show you the window resizing ;
:opaque resizes the window directly. (The box mode looks good on
laptops).
.TP
.B\-*verbose-move*\ \fIboolean\fP
If you want to have a small window that presents you the coordinates
of the the uper left corner of the window you are moving set it to
\fIt\fP otherwise \fInil\fP. The default value is \fIt\fP.
.TP
.B\-*verbose-resize*\ \fIboolean\fP
If you want to have a small window that presents you the sizes
of the window you are resizing set it to \fIt\fP otherwise
\fInil\fP. The default value is \fIt\fP.
.TP
.B\-*verbose-window-cycling*\ \fIboolean\fP
During window rotation, if set to \fIt\fP eclipse will display, in a
small window, the name of the focus window and its icon if
exist. Default value is \fIt\fP.
.TP
.B\-*cycle-icons-p*\ \fIboolean\fP
During window rotation, if set to \fIt\fP iconified window will be
included in the rotation. Default value is \fIt\fP. 
.TP
.B\-*warp-pointer-when-cycle*\ \fIboolean\fP
If set to \fIt\fP, the mouse pointer is warped around the upper-left
corner of the window that is virtually pointed to.  Otherwise, nothing 
happens. Useful when you work on a desktop with no overlapping
windows. The default value is \fIt\fP. 
.TP
.B\-*icon-hints*\ \fIboolean\fP
If set to \fIt\fP, eclipse will display small windows for representing
icons. If the iconified application provide a pixmap for its icon then
eclipse will display that pixmap in the icon window. Otherwise a part
of the application iconic name will be displayed. If you don't want
eclipse to display small window for icons just set it to
\fInil\fP. Default value is \fIt\fP. 
.TP
.B\-*screen-edge-resistant-p*\ \fIboolean\fP
Set it to \fInil\fP if you don't want to feel any resistance when
attempting to move a window outside the screen boundaries. Default
value is \fIt\fP.
.TP
.B\-*standard-window-edge-resistant-p*\ \fIboolean\fP
Set it to \fInil\fP if you don't want to feel any resistance on edges
of window(s) you are about to overlap. Default value is \fIt\fP.
.TP
.B\-*focus-when-window-cycle*\ \fIboolean\fP
If set to \fIt\fP the next window will be focused. Otherwise nothing 
happens. Useful when you work on a desktop with no overlapping
windows. The default value is \fIt\fP. 
.TP
.B\-*focus-new-mapped-window*\ \fIboolean\fP
If set to \fIt\fP newly mapped windows (or window comming from
iconified state) will ask for the focus. Otherwise nothing
happens. The default value is \fIt\fP.
.TP
.B\-*focus-type*\ \fIkeyword\fP (or \fB:none :on-click :when-switch\fP).

 \fB:none\fP focus change when enter/exit a window.
 \fB:on-click\fP focus change when click on a titlebar.
 \fB:when-switch\fP focus change when circulating window.

.TP
.B\-(setf decoration-theme) name &key free-old-theme-p\ \fIstring\fP
Change the theme. In general, for a list of all the themes that come
with \fIeclipse\fP, look in the directory called \fIthemes\fP. To use
any one of those present in that directory: just do \fB(setf
(decoration-theme) theme-name-string\fP. ("microGUI" for example).
But if the theme you want is not in the themes directory (because you
created it). Use the absolute pathname as the theme-name-string
argument of the \fB(setf decoration-theme)\fP for example:
"<$HOME>/tmp/my-new-theme". It also exists a "build-in" theme named
"no-decoration" that gives you undecorated windows. 
.TP
.B\-(setf number-of-virtual-screens) value\ \fIinteger\fP 
Indicate the number of virtual screens do you want.
.TP
.B\-(setf workspace-names) value\ \fIstring list\fP
To set the names of your workspaces. The length of the list of strings 
is suposed to be equal to the number of virtual screens.
.TP
.B\-(setf font-name) font-name-string\ \fIstring\fP
Sets the font used in title bar to display the application title.
The following pattern characters can be used for wildcard matching:
 \fB#\\*\fP Matches any sequence of zero or more characters. 
 \fB#\\?\fP  Matches any single character. 
For example, the pattern "T?mes Roman" matches the name "Times Roman"
but not the name "Thames Roman". However, the pattern "T*mes Roman"
matches both names. 
.TP
.B\-(setf maximize-modifier) value\ \fImodifier keyword\fP
Sets the modifier key that will be use to decide if we should maximize
the window (when pressing on the maximize button of a frame) with the
*maximize-fill*, if not down, or with (not *maximize-fill*), if down.

.SH THEMES
The image file format is \fBpnm P6\fP.
Here is a a decomposition of \fIeclipse\fP decorations :

       ________________________________
      |TL|MB         TOP       I M C|TR|
      |__|__________________________|__|
  __   ________________________________   __
 |TL| |TL|           TOP            |TR| |TR|
 |__| |__|__________________________|__| |__|
 |I | |  |                          |  | |I |
 |M | |  |                          |  | |M |
 |C | |  |                          |  | |C |
 |  | |  |                          |  | |  |
 |  | |  |                          |  | |  |
 |L | |L |                          |R | |R |
 |E | |E |                          |I | |I |
 |F | |F |                          |G | |G |
 |T | |T |                          |H | |H |
 |  | |  |                          |T | |T |
 |  | |  |                          |  | |  |
 |  | |  |                          |  | |  |
 |MB| |  |                          |  | |MB|
 |__| |__|__________________________|__| |__|
 |BL| |BL|          BOTTOM          |BR| |BR|
 |__| |__|__________________________|__| |__|
       ________________________________
      |BL|MB        BOTTOM     I M C|BR|
      |__|__________________________|__|

All part are optional.

Everything can be hilighted to reflect the focused state. You just
have to provide all necessary pnm files. For the push buttons MB menu
button, I iconify, M maximize, and C close, you can give four images
to reflect the button states: normal, focused, normal click and
focused click in that order.
The sizes of the buttons is not restricted. But they will be
considered as having the same height if your title bar is horizontal
(if vertical they will be considered as if they have the same width). 
.TP
.B\-size requirements\ 
TL height  = CENTER height = TR height
.br
BL height  = BOTTOM height = BR height

.TP
.B\-naming conventions\ 
Here is the corresponding table between file names and decoration parts: 


 part   |  file-name                     |  required

 TL        top-left.pnm                      no
 TR        top-right.pnm                     no
 BL        bottom-left.pnm                   no
 BR        bottom-right.pnm                  no
 TOP       top.pnm                           no
 BOTTOM    bottom.pnm                        no
 LEFT      left.pnm                          no
 RIGHT     right.pnm                         no

 I         iconify-normal.pnm                no 
           iconify-active.pnm                no
           iconify-normal-clicked.pnm        no
           iconify-active-clicked.pnm        no

 M         maximize-normal.pnm               no
           maximize-active.pnm               no
           maximize-normal-clicked.pnm       no
           maximize-active-clicked.pnm       no

 C         close-normal.pnm                  no
           close-active.pnm                  no
           close-normal-clicked.pnm          no
           close-active-clicked.pnm          no

MB         menu-button-normal.pnm            no
           menu-button-active.pnm            no
           menu-button-normal-clicked.pnm    no
           menu-button-active-clicked.pnm    no

.P
If you do not respect the recommendations above, there is no guarantee
concerning the resulting look and feel. 

.TP
.B\- Theme implemention notes\ 

.P
Themes are supposed to be defined in their own packages in a file
named theme.lisp.

Theme package definition must be supplied.

The naming convention for themes packages is:
<theme-name>-ECLIPSE-THEME. 
.br

Typically a theme package will export at least two symbols:
 - INITIALIZE-FRAME
 - REPAINT

Here is an example of how to define a theme (see our predefined
themes for more): 

.in +2
;;; -*- Mode: Lisp; Package: FOO-ECLIPSE-THEME -*-
.br
;;; GPL Disclaimer for example.
.br
;;; file theme.lisp
.br

(common-lisp:in-package :common-lisp-user)

(defpackage "FOO-ECLIPSE-THEME"
  (:use eclipse clx-ext common-lisp)
  (:size 10)
  (:export repaint initialize-frame)
  (:documentation
    "Foo theme decoration for the eclipse window manager. Written by ...")
  )

(in-package "FOO-ECLIPSE-THEME")

(define-theme "Foo"
.in +2
.br
((:default-style
.br
.in +3
(:title-bar-position :right)
.br
;; The parts to redraw on focus designator can be:
.br 
;; - :all to indicate that all the style parts should be
.br
;;    redrawn when focus change.
.br
;; - or null or a list of keywords as follows.
.br
(:parts-to-redraw-on-focus (:close :top-left))
.br
(;; The following are the frame parts for the default style.
.br
.in +1
(:top ("top"))
.br
(:close ("close-normal"
.br
.in +9
"close-active"
.br
"close-normal-clicked"
.br
"close-active-clicked"))
.br
.in -9
(:top-left ("top-left-inactive" "top-left-active"))
.br
(:top-right ("top-right"))
.br
(:bottom ("bottom"))
.br
(:bottom-right ("bottom-right"))
.br
(:bottom-left ("bottom-left"))
.br
(:right ("right"))
.br
(:left ("left"))))
.br
.in -3
(:transient-style
.in +3
.br
;; a background can be specified. It might be either:
.br
;; - an integer desining a color (e.g: an rgb value)
.br
;; - one of the :NONE :PARENT-RELATIVE keyword
.br
;; - a string that design a pnm file
.br
;; - a (xlib:make-color
.br
;;         :red <(real 0 1)>
.br
;;         :green <(real 0 1)>
.br
;;         :blue <(real 0 1)>)
.br
;;   form.
.br
(:background :parent-relative)
.br
.br
;; the rest of the definition
.br
)))
.in -6

(defmethod repaint ((widget title-bar) (name (eql "Foo")) (focus t))
  "handle repaint WHEN focused."
  ;; do some stuff
  )

(defmethod repaint ((widget title-bar) (name (eql "Foo")) (focus null))
  "handle repaint WHEN NOT focused."
  ;; do some stuff
  )

(defun bar ()
  ;; do what you want
  )

;;; end of theme.lisp
.in -2

.SH KEYSTROKES & MOUSESTROKES

All customisations and definitions take place in the config file.
To define a keyboard shortcut, use the macro \fBdefine-key-combo\fP.
You can define as well some mouse shortcut. To do so use the macro
\fBdefine-mouse-combo\fP. 

\fBdefine-key-combo\fP keystroke-name \fB&key\fP key modifiers
(default-modifiers-p t) fun 

\fBdefine-mouse-combo\fP mousestroke-name \fB&key\fP button modifiers 
(default-modifiers-p t) fun 

.TP
.B\-{key, mouse}stroke-name\ \fIkeyword\fP 
a keyword that names your stroke.

.TP
.B\-keys\ \fIkeyword\fP \fIlist\fP
a keyword list representing the keyboard key. (see \fIkeysymdef.lisp\fP)

.TP
.B\-button\ \fI(or number :any)\fP
a mouse button number or :any to indicate all buttons.

.TP
.B\-modifiers\ \fIkeyword\fP \fIlist\fP
a keyword list representing the keyword key, or a modifier-mask
(unsigned-byte 8). If it is a list, it can be of two differents forms:

.in +2
- '(:and :mod1 :mod2 ...)
.br
 to indicate a combination of modifiers.
.br
- '(:mo1 :mod2)
.br
 to indicate different possible modifiers.
.in -2

.br
Use \fB0\fP instead of \fInil\fP to indicate no modifier.
.br
Use \fB:any\fP to indicate any subset of modifier keys. 

.TP
.B\-default-modifiers-p\ \fIboolean\fP
default value is \fIt\fP. It allows you to define keystrokes with or
without \fB:CAPS-LOCK\fP and \fB:NUM-LOCK\fP.

.TP
.B\-fun\ \fIfunction\fP
a designator for a function or method of one argument. Its argument
will be the keyboard event that occurred. (i.e. key-press key-release
type) You can also use the macro \fBaction\fP to create your function:

.in +2
\fBaction\fP (\fB&rest\fP forms1) (\fB&rest\fP forms2)

It simply checks for keywords \fB:press\fP or \fB:release\fP to be the
\fIcar\fP of the forms and creates function as follows:

.in +4
(action (:press form1 form2) ())
.br
(action () (:press form1 form2))
.in +3
==> (lambda (event) 
.br 
.in +6
(typepcase event
.br
.in +2 
(key-press form1 form2)
.br
(key-release nil)))
.in -8
.in -3


(action
.br
.in +2
(:release form1 form2)
.br
(:press another-form))
.in -2
.in +3
==> (lambda (event) 
.br 
.in +6
(typepcase event
.br
.in +2 
(key-press another-form)
.br
(key-release form1 form2)))
.in -8
.in -3

.in -4
.in -2

.P
There exist some predefined keystrokes:

  - change window priority on the current desktop
    :switch-win-up  
    :switch-win-down

  - change of desktop
    :switch-screen-left
    :switch-screen-right

  - move the mouse cursor
    :move-right
    :move-left
    :move-up 
    :move-down

  - perform mouse click
    :left-click
    :middle-click
    :right-click

There exist some predefined mousestrokes:

  - resize a window 
    :resize-window

  - move a window
    :move-window

\fBKeystroke customisation example\fP:

\fB- predifined keystrokes:\fP
.in +1

(define-key-combo :switch-win-up
  :keys '(:TAB)
  :modifiers '(:ALT-LEFT))

(define-key-combo :switch-win-d 
  :keys '(:A)
  :modifiers '(:ALT-LEFT))

(define-key-combo :move-right
  :keys '(:RIGHT)
  :modifiers '(:and :ALT-LEFT :CONTROL-LEFT))
.in -1

\fB- a user defined one:\fP
.in +1


docs/changelog
-*- Mode: ChangeLog -*-
$Id: changelog,v 1.6 2004/02/12 23:30:23 ihatchondo Exp $
0.01 => 0.02

     Eclipse should now compile on every ANSI-compliant Common Lisp
     implementation.  This release actually looks quite usable.

     - sticky corrected.
     - window cycling corrected.
     - focus management is ICCCM complient. (thanks to Matthias Clasen)
     - pointer freezing corrections.
     - unexpected events during resize corrected.
     - unexpected events during move partially corrected.
     
     - gnome extended wm spec implemented and support added.
     - workspace names added.
     - button behavior changed : they are now "push buttons".
     - iconified windows now appears in the running menu like this: [ emacs ]
     - added some forgotten WITH-STANDARD-IO-SYNTAX. (thanks to Eric Marsden)
     - a mouse click in an application now puts it on top.
     - users can now define their own keystrokes. (see the man section)

     - complete change of keyboard handling. It works now without the 
       cmucl-hemlocks part. (thanks to Gilbert Bauman)
     - complete change of error-handling. (should work better)
     
     - some change in the autoconf script. Should now work correctly.
     - corrections in some parts of man and README.
       
0.02 => pre 0.03 (for James A. CRIPPEN)

     - bugs correction.
     - some change on sticky. should be 100% correct.
     - added X cursors.
     - added more Extended WM Spec complience.
     - update the gnome-manager.lisp (according to draft 1.2)
       - added utf8<->string converters (maybe not 100% correct)
       - all property that were suppose to be of type utf8 now are.
     - change "second button" menu behavior concerning the sticky window.
	They now apear in it.
     - change the way to add or remove desktop:
	now use (setf (number-of-virtual-screens) a-number)
     - added a new file virtual-screen.lisp which contain a sort of 
	generic implementation of the virtual-screen, and wm.lisp 
	now contain a specialisation of this class, named:
	eclipse-screens.
     - added two more predifined keystrokes for simulating the mouse buttons
        four and five (the wheel).
     - added some #. reader for avoiding some consing in some defconstant.

pre 0.03=>0.03

     - grid undraw failure during move in box type is now fixed.
     - added coordinates display during move and resize.
     - added opaque resize mode.
     - new theme interface (needs to be documented).
     - some new themes have been added (CoolClean, brushed-metal, Step).

0.03=>0.04

     - created an eclipse-internals package.
     - some bugs fix.
     - some functionalities added.
     - file splitting to reorganize the project a little:
     - wm.lisp and global.lisp had been splitted into:
	- widgets.lisp (base-widget, root, application, buttons, ...)
	- input.lisp (general event processing)
	- gestures.lisp (key-combo, cursor movement)
	- eclipse.lisp (Top level functions)
	- move-resize (moving and resizing master)
	- misc.lisp (...)
	- lib/clx-ext/clx-extensions.lisp
     - gnome-manager.lisp had been splitted into:
	- manager-commons.lisp
	- gnome-manager.lisp
	- netwm-manager.lisp
	- one directory added with one sub-directory: lib and lib/clx-ext
	- a new package clx-extensions in lib/clx-ext
	- some files had moved:
	  - image-reader.lisp into lib
	  - event.lisp ...... into lib/clx-ext
	  - cursor.lisp ..... ----------------
	  - cursordef.lisp .. ----------------
	  - keysysms.lisp ... ----------------
	  - keysymdef.lisp .. ----------------

     - commit Christian Lynbech patch:
       CMUCL seems to have tighten the handling of certain things so this
       patch will do three necessary things:

       - in system.lisp (compile-theme) remove path from :output-file name as
         this includes path specs from the input file.

       - in wm.lisp (menu-3-prcess) removes the &allow-other-keys which CMUCL
         18d chokes on (some previous versions seemed more lax, they also
	 accepted (lambda (&rest) ...) which 18d does not but you have gotten
	 rid of that).

      - in lib/image-reader.lisp (load-ppm) changes function to use two
        streams as there are many things you cannot do on a non-character
	string including calling `unread-char'.
	
0.04=>0.05

     - various bug fix. (with-state missing, undecore-application, some
       unhandle events during menu-3 loop, ...)
     - procede-decoration code is inside an unwind-protect.
     - support for focus-new-mapped-window
     - support for :_net_wm_state_maximized_{horz/vert},
       and :_net_wm_state_fullscreen.
     - fix typo parenthesis, and grid undraw in mouse stroke resize/move.
     - more netwm complience improvement (fullscreen, hidden, 
       icon-name, maximized)
     - Christian Lynbech patch for icon managment has been applied.
       "It provides an icon box, ie. a configurable place where icons are
        stacked. As part of this change, I have changed the default location
	of icons from (750,50) to (0,0). If the icon box overflows, icons will
	be created ayt (0,0). I have also decreased the default height from 60
	to 20 to minimize the space waste (if the application supplies an
	icon, it will of course take precedence).

	The patch allso adds different kinds of sorting options. Icons can be
	sorted by creation time (I have added a new slot in the `icon' class)
	or name or class (or whatever else you can think of).

	It still needs a few things, such as the possibility of disabling the
	icon box and other fill orders than :top-right (ie. filling from the
	top right corner of the box)."
     
     - Unrelated to the icon box thing, the patch also contains a set of
       application inspection function (see the end of "misc.lisp"), A
       disabling of the expression that removes the "Exit" entry in the root
       menu so that you can close Eclipse down and finally a new option,
       *close-display-p*, that regulates whether exiting Eclipse also kills

eclipse.lisp
;;; -*- Mode: Lisp; Package: ECLIPSE-INTERNALS -*-
;;; $Id: eclipse.lisp,v 1.26 2007/05/08 22:33:17 ihatchondo Exp $
;;;
;;; ECLIPSE. The Common Lisp Window Manager.
;;; Copyright (C) 2002 Iban HATCHONDO
;;; contact : hatchond@yahoo.fr
;;;
;;; This program is free software; you can redistribute it and/or
;;; modify it under the terms of the GNU General Public License
;;; as published by the Free Software Foundation.
;;;
;;; This program is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with this program; if not, write to the Free Software
;;; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA

(in-package :ECLIPSE-INTERNALS)

(defun load-config-file (pathname)
  (load pathname))

;;; Initializations and Main.

;; ICE & SM.
(defun sm-init (sm-conn dpy)
  "Sets the xsmp properties that are required by the protocols."
  (declare (type (or null string) dpy))
  (let ((id (format nil "--sm-client-id=~a" (sm-lib:sm-client-id sm-conn)))
	(display (format nil "--display=~a" dpy)))
    (ice-lib:post-request :set-properties sm-conn
      :properties
      (list (sm-lib:make-property
	     :name "CloneCommand"
	     :type "LISTofARRAY8"
	     :values (cons (sm-lib:string->array8 "eclipse")
			   (when dpy (sm-lib:strings->array8s display))))
	    (sm-lib:make-property
	     :name "Program"
	     :type "ARRAY8"
	     :values (sm-lib:strings->array8s "eclipse"))
	    (sm-lib:make-property
	     :name "RestartCommand"
	     :type "LISTofARRAY8"
	     :values (sm-lib:strings->array8s "eclipse" id))
	    (sm-lib:make-property
	     :name "UserID"
	     :type "ARRAY8"
	     :values (sm-lib:strings->array8s (get-username)))))))

(defun connect-to-session-manager (dpy-name &optional previous-id)
  "Try to connect us to the session manager. If connected set xsmp
   properties and returns the sm-connection instance."  
  (handler-case
      (let ((sm-conn (sm-lib:open-sm-connection :previous-id previous-id)))
	(sm-init sm-conn dpy-name)
	sm-conn)
    (error (condition) (format *error-output* "~&~A~&" condition))))

(defun handle-session-manager-request (sm-conn root-widget)
  "Handles xsmp requests. If a DIE request is received then invoke
   close-sm-connection and propagate the exit-eclipse condition."
  (handler-case
      (ice-lib:request-case (sm-conn :timeout 0)
	(sm-lib:save-yourself ()
          (ice-lib:post-request :save-yourself-done sm-conn :success-p t)
	  t)
	(sm-lib:die () (close-sm-connection root-widget :exit-p t) nil)
	(t t))
    (exit-eclipse (condition) (signal condition))
    (error (condition) (format *error-output* "~&~A~&" condition))))

(defun initialize-manager (display root-window)
  ;; ICCCM section 2.8
  (setf +xa-wm+ (format nil "WM_S~A" (xlib:display-display display)))
  (xlib:intern-atom display +xa-wm+)
  (let ((managing-since)
	(old-wm (xlib:selection-owner display +xa-wm+))
	(manager (xlib:create-window :parent root-window 
				     :override-redirect :on
				     :width 1 :height 1
				     :x 0 :y 0)))
    (when old-wm
      (setf (xlib:window-event-mask old-wm) '(:structure-notify)))

    ;; Get a valid timestamp.
    (with-event-mask (manager '(:property-change))
      (xlib:change-property manager :wm_name '(0) :string 8)
      (xlib:event-case (display :force-output-p t :discard-p t)
        (:property-notify (window time)
	  (when (xlib:window-equal window manager) (setf managing-since time)))
	(t nil)))

    ;; Ask for selection ownership, and wait for the old owner destruction.
    (setf (xlib:selection-owner display +xa-wm+ managing-since) manager)
    (when old-wm
      (xlib:event-case (display :force-output-p t :discard-p t :timeout 10)
        (:destroy-notify (window) (xlib:window-equal window old-wm))
	(t nil)))

    ;; Are we the selection owner after all ?
    (unless (xlib:window-equal manager (xlib:selection-owner display +xa-wm+))
      (error "ICCCM Error: failed to aquire selection ownership~%"))

    ;; Check if a non ICCCM complient window manager is not running.
    (handler-case
	(progn
	  (setf (xlib:window-event-mask root-window) +root-event-mask+)
	  (xlib:display-finish-output display))
      (error () (error "Redirect error: another WM is running~%")))

    ;; Notify all the other X clients of the new manager.
    (xlib:send-event root-window :client-message '(:structure-notify)
      :window root-window
      :type :MANAGER
      :format 32
      :data (list managing-since 
		  (xlib:find-atom display +xa-wm+)
		  (xlib:window-id manager)))
    (setf (xlib:window-event-mask manager) '(:property-change))
    (make-instance 'standard-property-holder :window manager)))

(defun init-gnome-compliance (display window manager)
  (gnome:intern-gnome-atom display)
  (netwm:intern-atoms display)
  (let ((first-desknum (current-vscreen window))
	(nb-vs (number-of-virtual-screens window))
	(srcw (screen-width)) (srch (screen-height)))
    (xlib:with-server-grabbed (display)
      (delete-properties window +netwm-protocol+)
      (unless (< -1 first-desknum nb-vs) (setf first-desknum 0))
      (setf (gnome:win-protocols window) +gnome-protocols+
            (gnome:win-supporting-wm-check manager) manager
            (gnome:win-supporting-wm-check window) manager
            (gnome:win-workspace-count window) nb-vs
            (gnome:win-workspace window) first-desknum)

      (setf (netwm:net-supported window) +netwm-protocol+
            (netwm:net-supporting-wm-check window) manager
            (netwm:net-supporting-wm-check manager) manager
            (netwm:net-wm-name manager) "eclipse"
            (netwm:net-number-of-desktops window) nb-vs
            (netwm:net-current-desktop window) first-desknum
            (netwm:net-desktop-viewport window) (make-viewport-property nb-vs)
            (netwm:net-desktop-geometry window) (list srcw srch)
            (netwm:net-workarea window) (make-list nb-vs
                                            :initial-element
                                            (manager-commons:make-geometry-hint
                                             :x 0 :y 0 :width srcw :height srch))

	  ))))

(defun initialize (display-specification sm-client-id)
  (multiple-value-bind (display screen)
      (open-clx-display display-specification)
    (let* ((colormap (xlib:screen-default-colormap screen))
	   (root-window (xlib:screen-root screen))
	   (manager (initialize-manager display root-window))
	   (menu-font (xlib:open-font display "fixed")))
      (setf *display* display)
      ;; Specific for X display
      (setf (xlib:display-error-handler display) #'default-handler
	    (xlib:display-after-function display) #'xlib:display-force-output)
      (setf *root* (make-instance 'root :window root-window :manager manager)
	    *root-window* root-window
	    (root-default-cursor *root*) (get-x-cursor display :xc_left_ptr)
	    (root-sm-conn *root*) (connect-to-session-manager
				      display-specification sm-client-id))
      ;; init all gnome properties on root.
      (init-gnome-compliance display root-window (widget-window manager))
      (ppm:initialize colormap)
      ;; Eclipse globals vars.
      (setf *black* (xlib:screen-black-pixel screen)
	    *white* (xlib:screen-white-pixel screen)
	    *background1* (xlib:alloc-color colormap *menu-color*)
	    *background2* (xlib:alloc-color colormap *menu-hilighted-color*)
	    *cursor-2* (get-x-cursor display :xc_fleur)
	    *gctxt* (xlib:create-gcontext :drawable root-window :font menu-font)
	    *max-char-width* (xlib:max-char-width menu-font)
	    *gcontext* (xlib:create-gcontext
			  :drawable root-window
			  :foreground *white* :background *black*
			  :fill-style :solid  :line-style :solid
			  :line-width 1 :exposures :OFF))
      ;; load personal configuration file, or the default one.
      (labels ((load-if (f) (and (probe-file f) (load-config-file f))))
	(or (load-if (home-subdirectory cl-user::*eclipse-initfile*))
	    (load-if (eclipse-path "eclipserc"))
	    (error "Unable to read a configuration file.~%")))
      (setf (xlib:window-cursor root-window) (root-default-cursor *root*))
      (setf (slot-value *root* 'gcontext) *gcontext*)
      (unless (xlib:gcontext-font *gcontext*)
	(setf (font-name) +default-font-name+))
      (unless (root-decoration-theme *root*)
	(setf (decoration-theme) "microGUI")))))

(defun eclipse (&key display sm-client-id die-on-init-error activate-log)
  "Starts the Eclipse window manager.
    - :display (or null string): if given it is expected it respects the 
      standard X DISPLAY specification: hostname:displaynumber.screennumber
      (for more about the display string format see the X manual page).
      If NIL then the DISPLAY environment variable will be used.
    - :sm-client-id (or null string): if Eclipse is restarted from a previous
      session, should contain the previous client-id of that previous session.
      If :sm-client-id is specified, but is determined to be invalid by the
      session manager, we will re-register the client with a sm-client-id set
      to NIL. If the client is first joining the session :sm-client-id can be
      NIL (default) or the empty string.
    - :die-on-init-error (boolean): indicates if Eclipse should prevent
      from debugger or not during initialisation phase.
    - :activate-log (boolean): indicates if errors should be logged.
      If T then errors will be logged in a file named: eclipse-yyyy-mm-dd.log
   If neither the DISPLAY environment variable is defined nor the :display
   argument is defined then Eclipse will not be able to starts."
  (declare (type (or null string) display sm-client-id))
  (declare (type boolean activate-log die-on-init-error))
  (when *display* 
    (xlib:close-display *display*)
    (setf *display* nil))
  (if die-on-init-error
      (handler-case (initialize display sm-client-id)
	(error (condition)
	  (format *error-output* "~A~%" condition)
	  (%quit%)))
      (initialize display sm-client-id))
  (when activate-log
    (init-log-file))

  ;; This is for releasing any previous pointer grab.
  (ignore-errors
    (grab-root-pointer)
    (xlib:display-finish-output *display*))
  (xlib:ungrab-pointer *display*)

  ;; Create a socket connection to communicate with the window manager.
  ;; Works only for CMUCL -x86- (unless you compile this mp package).
  #+:mp (progn
	  (setf mp::*idle-process* mp::*initial-process*)
	  (mp::start-lisp-connection-listener :port 6789 :password "clara"))

  (unwind-protect
      (catch 'end
	(handler-bind ((end-of-file #'handle-end-of-file-condition))
	  (eclipse-internal-loop)))
    (progn
      (ignore-errors (xlib:close-display *display*))
      (format t "Eclipse exited. Bye.~%")
      (%quit%))))
